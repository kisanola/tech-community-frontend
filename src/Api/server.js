import axios from 'axios';

const token = localStorage.getItem('token') || null;
const server = axios.create({
  baseURL: process.env.REACT_APP_BACKEND_URL,
  headers: {
    Authorization: token,
    'Content-Type': 'application/json',
  },
});

server.interceptors.request.use(
  async config => {
    const token = await localStorage.getItem('token');
    if (token) config.headers['Authorization'] = token;
    return config;
  },
  error => {
    return Promise.reject(error);
  },
);

export default server;
