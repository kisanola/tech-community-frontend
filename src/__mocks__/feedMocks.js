export default {
  author: { username: 'name', picture: 'https://image.png' },
  userType: 'organization',
  image: 'https://image.png',
  description: 'description',
  likesCount: 1,
  createdAt: '2019-09-22',
};
