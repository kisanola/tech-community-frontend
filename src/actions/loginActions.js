import axios from 'axios';
import {
  LOGIN_ERROR,
  LOGIN_SUCCESS,
  LOGIN_STARTED,
  LOGIN_INPUT_CHANGED,
} from '../actionTypes/loginTypes';

/**
 * Updates the form inputs
 * @param {*} attribute
 * @return {object} action
 */
export const formInputChanged = attribute => ({
  type: LOGIN_INPUT_CHANGED,
  payload: attribute,
});

/**
 * Triggers an action
 * when the process starts
 * @return {object} action
 */
export const loginStarted = () => ({
  type: LOGIN_STARTED,
});

/**
 * Triggers an action
 * when the process succeed
 * @param {object} user
 * @return {object} action
 */
export const loginSuccess = user => ({
  type: LOGIN_SUCCESS,
  payload: user,
});

/**
 * Triggers an action
 * when the process fails
 * @param {object} error
 * @return {object} action
 */

export const loginError = error => ({
  type: LOGIN_ERROR,
  payload: error,
});

/**
 * Login the user
 * @param {*} { username, email, password }
 * @return {object} response
 */
export const login = ({ email, password }) => async dispatch => {
  dispatch(loginStarted());

  try {
    const res = await axios.post(
      `${process.env.REACT_APP_BACKEND_URL}/api/v1/auth/login`,
      {
        username: email,
        password,
      },
      {
        headers: {
          'Content-Type': 'application/json',
        },
      },
    );

    const { token, user } = res.data;

    localStorage.setItem('token', token);

    dispatch(loginSuccess(user));
    return true;
  } catch (e) {
    if (e.response) {
      if (e.response.data.errors) {
        dispatch(loginError(e.response.data.errors[0].message));
        return false;
      }
      dispatch(loginError(e.response.data.message));
      return false;
    }
    dispatch(loginError('Please check your internet connection'));
    return false;
  }
};
