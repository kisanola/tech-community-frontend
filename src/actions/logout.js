import { LOGOUT_SUCCESS } from '../actionTypes/currentUserTypes';

export const setIsLoggedOut = () => dispatch => {
  localStorage.removeItem('username');
  localStorage.removeItem('token');
  return dispatch({
    type: LOGOUT_SUCCESS,
    payload: true,
  });
};
