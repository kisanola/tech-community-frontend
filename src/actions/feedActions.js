import server from '../Api/server';

import * as types from '../actionTypes/feed';

import { FEED_LIMIT } from '../constants';

/**
 * Toggles the loading state
 * @return {object} action
 */
export const toggleLoading = (state) => ({
  type: types.TOOGLE_LOADING,
  payload: state,
});

/**
 * Set feed
 * @return {object} action
 */
export const setFeed = (payload) => ({
  type: types.SET_FEED,
  payload,
});

/**
 * Set feed organization
 * @return {object} action
 */
export const setFeedOrganizations = (payload) => ({
  type: types.SET_FEED_ORGANIZATIONS,
  payload,
});

/**
 * Get feed
 * @return {object} response
 */
export const getFeed = ({
  limit,
  itemsLength,
  category = 'general',
  search = '',
}) => async (dispatch) => {
  if (limit > itemsLength + FEED_LIMIT) return;

  dispatch(toggleLoading(true));

  try {
    const {
      data: { feed },
    } = await server.get(
      `/api/v1/feed?offset=${itemsLength}&limit=${FEED_LIMIT}&category=${category}&search=${search}`,
    );

    dispatch(setFeed(feed));
    dispatch(toggleLoading(false));
  } catch (e) {
    dispatch(toggleLoading(false));
  }
};


/**
 * Get feed
 * @return {object} response
 */
export const getFeedOrganizations = () => async (dispatch) => {
  const res = await server.get(
    '/api/v1/organizations?limit=5',
  );
  if (res) {
    const {
      data: { organizations },
    } = res;
    dispatch(setFeedOrganizations(organizations.docs || []));
  }
};
