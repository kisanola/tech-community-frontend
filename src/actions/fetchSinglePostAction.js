import server from '../Api/server';

import * as types from '../actionTypes/singlePostTypes';

/**
 * Toggles the loading state
 * @return {object} action
 */
export const toggleLoading = (state) => ({
  type: types.TOGGLE_LOADING,
  payload: state,
});

/**
 * Set feed
 * @return {object} action
 */
export const fetchSuccess = (post) => ({
  type: types.FETCH_SUCCESS,
  payload: post,
});

/**
 * Get feed
 * @return {object} response
 */
export default (slug) => async (dispatch) => {
  dispatch(toggleLoading(true));

  try {
    const {
      data: { post },
    } = await server.get(
      `/posts/${slug}`,
    );

    dispatch(fetchSuccess(post));
    dispatch(toggleLoading(false));
  } catch (e) {
    dispatch(toggleLoading(false));
  }
};
