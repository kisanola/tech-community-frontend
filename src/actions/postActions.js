import axios from 'axios';
import server from '../Api/server';
import {
    POST_ERROR,
    POST_STARTED,
    POST_SUCCESS
} from '../actionTypes/postTypes';

const {
    REACT_APP_CLOUDINARY_URL_IMAGE,
    REACT_APP_CLOUDINARY_PRESET,
} = process.env;

/**
 * Triggers an action
 * when the process starts
 * @return {object} action
 */
const postStarted = () => ({ type: POST_STARTED })

/**
 * Triggers an action
 * when the process succeed
 * @param {object} post
 * @return {object} action
 */
const postSuccess = (post) => ({
    type: POST_SUCCESS,
    payload: post
})

/**
 * Triggers an action
 * when the process fails
 * @param {object} error
 * @return {object} action
 */
const postError = (error) => ({
    type: POST_ERROR,
    payload: error
})

const uploadImage = async file => {
    const formData = new FormData();
    formData.append('file', file);
    formData.append('upload_preset', REACT_APP_CLOUDINARY_PRESET);

    try {
        const res = await axios({
            url: REACT_APP_CLOUDINARY_URL_IMAGE,
            method: 'POST',
            header: {
                'Content-Type': 'application/x-ww-form-urlencoded',
            },
            data: formData,
        })
        return res.data.secure_url

    } catch (error) {
        throw error;
    }
};


/**
 * post
 * @param {*} { username, email, password }
 * @return {object} response
 */
const post = (data) => async dispatch => {
    dispatch(postStarted());

    try {
        const payload = {
            description: data.value,
            title: 'this is a wrong title'
        };

        if (data.image) payload.image = await uploadImage(data.image);

        const res = await server.post(`/api/v1/posts`, payload);

        const { post } = res.data;
        dispatch(postSuccess(post));
        return true;
    } catch (e) {
        if (e.response) {
            if (e.response.data.errors) {
                dispatch(postError('The post should have at least 50 characters long'));
                return false;
            }
            dispatch(postError(e.response.data.message));
            return false;
        }
        dispatch(postError('Please check your internet connection'));
        return false;
    }
};

export default post;

