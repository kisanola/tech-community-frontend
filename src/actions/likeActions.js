import server from '../Api/server';

import { toggleLoading } from './feedActions';

/**
 * Like a post
 * @return {object} action
 */
export const like = (postSlug) => async (dispatch) => {
  dispatch(toggleLoading(true));

  try {
    await server.post(`/posts/${postSlug}/like`);
    dispatch(toggleLoading(false));
  } catch (e) {
    dispatch(toggleLoading(false));
  }
};

/**
 * Unlike a post
 * @return {object} action
 */
export const unlike = (postSlug) => async (dispatch) => {
  dispatch(toggleLoading(true));

  try {
    await server.delete(`/posts/${postSlug}/like`);
    dispatch(toggleLoading(false));
  } catch (e) {
    dispatch(toggleLoading(false));
  }
};
