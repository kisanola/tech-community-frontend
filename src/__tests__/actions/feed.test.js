import server from '../../Api/server';
import { getFeed } from '../../actions/feedActions';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import feedMocks from '../../__mocks__/feedMocks';

const mockStore = configureMockStore([thunk]);

describe('Login actions', () => {
  let store;

  beforeEach(() => {
    store = mockStore({});
  });

  test('Should return when the limit is greater than the items length', () => {
    store.dispatch(getFeed(40, 0)).then(() => {
      expect(store.getActions()).not.toMatchSnapshot();
    });
  });

  test('Should call the `getFeed` action with success', () => {
    jest.spyOn(server, 'get').mockResolvedValue({
      data: { feed: [feedMocks] },
    });

    store.dispatch(getFeed(0, 0)).then(() => {
      expect(store.getActions()).toMatchSnapshot();
    });
  });

  test('Should call the `getFeed` action with success', () => {
    jest.spyOn(server, 'get').mockRejectedValue({
      data: { feed: [feedMocks] },
    });

    store.dispatch(getFeed(0, 0)).then(() => {
      expect(store.getActions()).toMatchSnapshot();
    });
  });
});
