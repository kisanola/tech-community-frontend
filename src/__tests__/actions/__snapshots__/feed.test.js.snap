// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login actions Should call the \`getFeed\` action with success 1`] = `
Array [
  Object {
    "payload": true,
    "type": "TOOGLE_LOADING",
  },
  Object {
    "payload": Array [
      Object {
        "author": Object {
          "picture": "https://image.png",
          "username": "name",
        },
        "createdAt": "2019-09-22",
        "description": "description",
        "image": "https://image.png",
        "likesCount": 1,
        "userType": "organization",
      },
    ],
    "type": "SET_FEED",
  },
  Object {
    "payload": false,
    "type": "TOOGLE_LOADING",
  },
]
`;

exports[`Login actions Should call the \`getFeed\` action with success 2`] = `
Array [
  Object {
    "payload": true,
    "type": "TOOGLE_LOADING",
  },
  Object {
    "payload": false,
    "type": "TOOGLE_LOADING",
  },
]
`;
